Add code signing preview page (#230)
shorebirdtech/docs: src/content/docs/patch-signing.mdx

---
title: Patch Signing (preview)
description: How to sign patches
---

{/* cspell:words outform pubout */}

:::note
We've shipped this preview of patch signing to unblock certain customer
evaluations. While we're confident that this preview adds a useful layer of
security, it does not yet have all the expected integrations with cloud signing,

> keychains, etc. that a final product would expect to have. We also do not yet

Reviewer: ```suggestion
keychains, etc, that a final product would expect to have. We also do not yet
```

Author: "etc." should be followed by a period as it is an abbreviation.

manage signing keys for you or support many key formats or signature types.
Signed patches are safe and ready for use, but may have rough edges or missing
conveniences. Any customer should feel welcome to use this preview in

> production, and is encouraged to file feedback for any rough edges you may find.

Reviewer: ```suggestion
production, and is encouraged to file feedback for any bumps you may find.
```
Or something else, just so we don't repeat `rough edges`

Author: Done

:::

In addition to our default security measures, Shorebird also provides optional
patch signing.

Patch signing allows developers to cryptographically sign patch updates with
their own keys. This ensures that no one (including Shorebird) can change the

> content of your application without your private cryptographic keys.

Reviewer: ```suggestion
content of your patch without your private cryptographic keys.
```

Author: Done

Signing works in two parts. First, `shorebird release` commands can take an

> optional `--public-key-path` argument to embed a your public key in your

Reviewer: ```suggestion
optional `--public-key-path` argument to embed a public key in your
```

Author: Done

released application. The Shorebird updater will enforce that only patches
signed with a corresponding private key will be allowed to load for applications
that include a public key.

Second, when you build your patch with `shorebird patch`, you can pass
`--private-key-path` to have Shorebird sign your patch with your private key.

> This is required if you created your release with a public key included.

Reviewer: We now require the `public key` too

Author: Good catch, added.

There are no required changes to your code and you can add or remove this
signing requirement at any time by simply making a new release of your
application.

## Adding patch signing to your application

To start, you will need an RSA key pair. Shorebird has only tested with a
limited set of signing algorithms at this time. Please contact us if you have

> other requirements—we'd be happy to work with you to support your needs.

Reviewer: ```suggestion
other requirements, we'd be happy to work with you to support your needs.
```

Author: An em-dash is more correct here

The `shorebird` tool expects to be able to read private and public keys
from `.pem` files on disk. If you need other ways of accessing your key material

> (or signing via a cloud signing service) please contact us. We'd be happy to

Reviewer: I feel this paragraph kinda steps into what the previous one already said, about we being happy to support other needs.

If you do not already have an RSA key pair you'd like to use, you can generate a
pair with `openssl`:

```sh
# Generate a key pair
openssl genrsa -out private.pem 2048

# Extract the public key
openssl rsa -in private.pem -outform PEM -pubout -out public.pem
```

The above operation will generate a public/private key pair in `private.pem` and
a public key in `public.pem`. The file containing the private key should be
stored securely and kept secret. While the private key is not itself sufficient
to make an update to your application (someone would also need access to your
Shorebird credentials), it should not be checked into public source control.

### Create a release containing the public key

To create a release that requires signed patches, run the following command:

```sh
shorebird release android --public-key-path /path/to/public.pem
```

This will include the public key in the release artifact produced by `shorebird
release` and cause the released app to require signed patches.

:::note
We're using Android to demo, but this will work on any platform Shorebird
supports.
:::

### Create a signed patch

To create a signed patch, run the following command:

```sh

> shorebird patch android --private-key-path /path/to/private.pem

Reviewer: ```suggestion
shorebird patch android --private-key-path /path/to/private.pem --public-key-path /path/to/public.pem
```

Author: Done

```

> This tells shorebird to sign the patch with the private key you provide.

Reviewer: ```suggestion
This tells shorebird to sign the patch with the key pair you provided.
```

:::note
Storing private keys on disk is not best-practice, and we recognize
this. This was built as a demonstration of the system, we look forward to
working with customers to integrate Shorebird to using the signing services and
key storage methods they are used to.
:::

### Test it out

You can verify that the patch is properly signed using `shorebird preview`. On

> the first launch, you should see something like the following in your app logs:

Reviewer: Are the logs missing?

Author: Ha, good catch. The Google-Doc-to-markdown converter I used turns out to not be very good 😬

Author: Fixed

If you close and relaunch your app, you should see this message telling you that

> the patch's signature was verified:

Reviewer: Same here

Author: Fixed

reason, Shorebird will not load it, and will instead use any previously
installed and verifiable patch (if there is one) or the unpatched release
version of your app.

If you'd like to test the missing signature behavior, you can create a patch
without a private key and notice that `shorebird preview` rejects it. Similarly,
you can create a patch with a different private key to do the same.

## Fallback Behavior

Releases that contain a public key will reject all unsigned patches.
If a patch is missing a signature or its signature is invalid, Shorebird will
reject this patch at boot time. It will instead boot from the last known good
patch (if still on disk) or the release build of the app. This will not cause
your app to crash.

If for any reason you were to lose your private key, there is no way to create a
patch for an application containing the corresponding public key. Even Shorebird
is not able to create a patch for your application without your private key.
In such a case, you would need to make and distribute a new release of your
application to send patches to it.

## Trade Offs

The primary trade-off of using patch signing is complexity. Shorebird does not

> yet offer automatic key management, so customers you to create and manage their

Reviewer: ```suggestion
yet offer automatic key management, so customers are to create and manage their
```

Author: Fixed
